### PR TITLE
[flutter_tools] Cleanup of native asset related code (removes around 50% of the native asset related code)

### PR DESCRIPTION
**Pull request by mkustermann**
tl;dr Removes 50% (>1650 locs) of native asset related code in `packages/flutter_tools`

Before this PR the invocation of dart build/link/dry-run was implemented per OS. This lead to very large code duplication of almost identical, but sligthly different code. It also led to similarly duplicated test code.

Almost the entire dart build/link/dry-run implementation is identical across OSes. There's small variations:

  - configuration of the build (e.g. android/macos/ios version, ios sdk, ...)
  - determining target locations & copying the final shared libraries

This PR unifies the implementation by reducing the code to basically two main functions:

  * `runFlutterSpecificDartBuild` which is responsible for
    - obtain flutter configuration
    - perform dart build (& link)
    - determine target location & install binaries

  * `runFlutterSpecificDartDryRunOnPlatforms` which is responsible for a similar (but not same):
    - obtain flutter configuration
    - perform dart dry run
    - determine target location

these two functions will call out to helpers for the OS specific functionality:

  * `_assetTargetLocationsForOS` for determining the location of the code assets

  * `_copyNativeCodeAssetsForOS` for copying the code assets (and possibly overriting the install name, etc)

=> Since we get rid of the code duplication across OSes and have only a single code path for the build/link/dry-run, we can also remove the duplicated tests that were pretty much identical across OSes.

We also harden the building code by adding asserts, e.g.

  * the dry fun functionality should never be used by `flutter test`

  * the `build/native_assets/<os>/native_assets.yaml` should only be used by `flutter test` and the dry-run of `flutter run`

=> We change the tests to also comply with these invariants (so the tests are not testing things that cannot happen in reality)

We also rename `{,Flutter}NativeAssetsBuildRunner` to disambiguate it from the `package:native_asset_builder`'s `NativeAssetsBuildRunner`.

We also reorganize the main code to make it readable from top-down and make members private where they can be.

### CODE REVIEW

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 1]
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
...
+    FileSystem fileSystem,
+    Map<NativeCodeAssetImpl, KernelAsset> assetTargetLocations,
+    String? codesignIdentity,
+    bool flutterTester) async {
+  final List<NativeCodeAssetImpl> codeAssets =
+      assetTargetLocations.keys.toList();
+  switch (targetOS) {
+    case OS.windows:
+    case OS.linux:
+      await _copyNativeCodeAssetsToBundleOnWindowsLinux(
```
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:686]: Maybe assert that `codesignIdentity` is null. Since we don't do code signing. If we ever add code signing, we should not forget to use it here.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 2]
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
   return projectUri.resolve('$buildDir/native_assets/$os/');
 }
 
-class HotRunnerNativeAssetsBuilderImpl implements HotRunnerNativeAssetsBuilder {
-  const HotRunnerNativeAssetsBuilderImpl();
+/// With `flutter run -d all` we need a place to store the native assets
```
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:568]: Nit: The diff is larger than it needs to be due to reorderings. This makes reviewing harder and redirects questions to you via git blame. If you feel strongly about the order of code, fine be me.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 3]
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
...
-      fileSystem,
-      globals.logger,
-    );
-    return dryRunNativeAssets(
-      projectUri: projectUri,
-      fileSystem: fileSystem,
-      buildRunner: buildRunner,
-      flutterDevices: flutterDevices,
-    );
+Map<NativeCodeAssetImpl, KernelAsset> _assetTargetLocationsWindowsLinux(
```
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:575]: The previous reviewers preferred not having OS-specific things in this OS-agnostic file. Please rename back to `_assetTargetLocationsSingleArchitecture` to comply with their wishes.

And ditto for code moving.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 4]
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
...
-/// Needed for `flutter run -d all`.
-Future<Uri?> dryRunNativeAssetsMultipleOSes({
-  required NativeAssetsBuildRunner buildRunner,
-  required Uri projectUri,
-  required FileSystem fileSystem,
-  required Iterable<build_info.TargetPlatform> targetPlatforms,
-}) async {
-  if (await nativeBuildRequired(buildRunner)) {
-    return null;
+Future<void> _copyNativeCodeAssetsToBundleOnWindowsLinux(
```
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:726]: Ditto, avoid using OSspecific names in this file.

Ditto, shows up as diff even though didn't change.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 5]
```diff
@@ -410,417 +565,363 @@ Uri nativeAssetsBuildUri(Uri projectUri, OS os) {
...
-    fileSystem,
-  );
-  return nativeAssetsUri;
-}
+  final CCompilerConfigImpl cCompilerConfig = targetOS == OS.android
+      ? await buildRunner.ndkCCompilerConfigImpl
+      : await buildRunner.cCompilerConfig;
+  final IOSSdkImpl? targetIOSSdk =
+      targetOS == OS.iOS ? getIOSSdk(environmentType!) : null;
+  // TODO(dcharkes): Fetch minimum iOS version from somewhere. https://github.com/flutter/flutter/issues/145104
```
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:792]: OS-specific things should live in the OS-specific files. Maybe move helper functions there (also for targetIOSVersion and targetMacOSVersion) and leave a comment on the `buildRunner.build` invocation it contains the config for all OSes. And then on the helper functions that they return null for the irrelevant OSes?

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart [hunk 6]
```diff
@@ -99,13 +100,13 @@ Future<void> setInstallNamesDylib(
...
 ) async {
-   final ProcessResult setInstallNamesResult = await globals.processManager.run(
+  final ProcessResult setInstallNamesResult = await globals.processManager.run(
     <String>[
       'install_name_tool',
       '-id',
       newInstallName,
-      for (final MapEntry<String, String> entry in oldToNewInstallNames.entries)
-        ...<String>['-change', entry.key, entry.value],
+      for (final MapEntry<String, String> entry in oldToNewInstallNames
```
Comment (dcharkes) [inline packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:108]: I hate to do this. But it looks like the formatter ran and made a bunch of whitespace changes. (Also in other files.)

I will refer to the Flutter team on if we care about that in the days before turning the formatter on here in the codebase.

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart [hunk 7]
```diff
@@ -135,6 +135,7 @@ void main() {
               const Barrier('Application finished.'),
             ],
             logging: false,
+            debug: true,
```
Comment (dcharkes) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:138]: Revert this?

**Review by dcharkes — APPROVED** [review]
LGTM!

Thanks @mkustermann! 🚀 

I'll refer to the Flutter team on unrelated code having been formatted by the formatter.

**Comment by dcharkes** [comment]
Probably @christopherfujino or @andrewkolos should review this. (Or @bkonyi.)

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 1 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:686]: Done - though I've had an assert in another place already.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 3 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:575]: If we have `_assetTargetLocations{Android,IOS,MacOS}` a reader of this code would naturally look for `_assetTargetLocations{Linux,Windows}` and cannot find them - which is the first confusing.

If we name it `_assetTargetLocationsSingleArchitecture`, then one could think that `_assetTargetLocationsAndroid`  means it's always multiple architectures, but it's not necessarily the case.

The difference between these functions isn't whether it's a single architecture or multiple but which operating system they deal with. So I think it makes more sense to have this naming.

@dcharkes If you feel strongly about this, I can rename it back, but I think it's worse for a reader of this code.

Comment (dcharkes) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:575]: > @dcharkes If you feel strongly about this, I can rename it back, but I think it's worse for a reader of this code.

I don't, I had `LinuxWindows` in my initial PRs as well before the reviewers on those PRs requested changes.

Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:575]: Then I'll leave it as it is.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 4 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:726]: It makes consistent naming with all the other `_copy*` methods.

I've tried to mimize the diff now as much as I could.

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart (hunk 6 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:108]: I've tried to manually revert the auto-formatted lines.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 5 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:792]: Done.

**Comment by mkustermann** [comment]
@dcharkes I've rebased, tried to align old/new code to make diff smaller, tried to un-auto-format code. Please let me know if there's something else.

@bkonyi This is mainly a refactoring of native asset code (for which I'd consider @dcharkes the most knowledgeable person / owner) - do you also want to review this?

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart [hunk 8]
```diff
@@ -271,7 +271,7 @@ void main() {
             'build',
             buildSubcommand,
             if (buildSubcommand == 'ios') '--no-codesign',
-            '-v', // Requires verbose mode for error.
+            '-v' // Requires verbose mode for error.
```
Comment (dcharkes) [inline packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart:274]: formatting merge change?

**Review by dcharkes — APPROVED** [review]
lgtm from my side!

File: packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart (hunk 8 above)
Comment (mkustermann) [reply packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart]: Done.

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart [hunk 9]
```diff
@@ -99,13 +99,13 @@ Future<void> setInstallNamesDylib(
...
```
Comment (bkonyi) [inline packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:107]: Ubernit: consider using an object destructuring pattern to pull out the key and value.

File: packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart [hunk 10]
```diff
@@ -287,12 +123,16 @@ void main() {
...
         environment.projectDir.childFile('.dart_tool/package_config.json');
+    final Uri nonFlutterTesterAssetUri = environment.buildDir.childFile('native_assets.yaml').uri;
     await packageConfig.create(recursive: true);
-    await buildNativeAssetsAndroid(
-      androidArchs: <AndroidArch>[AndroidArch.x86_64],
-      targetAndroidNdkApi: 21,
+    await runFlutterSpecificDartBuild(
+      environmentDefines: <String, String>{
+        kBuildMode: BuildMode.debug.cliName,
+        kMinSdkVersion: 21.toString(),
```
Comment (bkonyi) [inline packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart:131]: Can this be a constant somewhere?

File: packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart [hunk 11]
```diff
@@ -308,16 +148,19 @@ void main() {
...
     await packageConfig.parent.create();
     await packageConfig.create();
     expect(
-      () => buildNativeAssetsAndroid(
-        androidArchs: <AndroidArch>[AndroidArch.arm64_v8a],
-        targetAndroidNdkApi: 21,
+      () => runFlutterSpecificDartBuild(
+        environmentDefines: <String, String>{
+          kBuildMode: BuildMode.debug.cliName,
+          kMinSdkVersion: 21.toString(),
```
Comment (bkonyi) [inline packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart:158]: Ditto

File: packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart [hunk 12]
```diff
@@ -355,22 +186,25 @@ void main() {
...
-            ),
+        ),
       );
-      await buildNativeAssetsIOS(
-        darwinArchs: <DarwinArch>[DarwinArch.arm64, DarwinArch.x86_64],
-        environmentType: EnvironmentType.simulator,
+      await runFlutterSpecificDartBuild(
+        environmentDefines: <String, String>{
+          kBuildMode: buildMode.cliName,
+          kSdkRoot: '.../iPhone Simulator',
```
Comment (bkonyi) [inline packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:194]: I'm not super familiar with this code, but is this supposed to be a path?  What is this value for?

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 13]
```diff
@@ -30,11 +29,159 @@ import 'macos/native_assets.dart';
 import 'macos/native_assets_host.dart';
 import 'windows/native_assets.dart';
 
+/// The assets produced by a Dart build and the dependencies of those assets.
+///
+/// If any of the dependencies change, then the Dart build should be performed
+/// again.
+final class DartBuildResult {
+  const DartBuildResult(this.codeAssets, this.dependencies);
```
Comment (bkonyi) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:37]: Nit: prefer required named arguments.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 14]
```diff
@@ -30,11 +29,159 @@ import 'macos/native_assets.dart';
...
+  const DartBuildResult.empty()
+      : codeAssets = const <NativeCodeAssetImpl>[],
+        dependencies = const <Uri>[];
+
+  final List<NativeCodeAssetImpl> codeAssets;
+  final List<Uri> dependencies;
+}
+
+/// Invokes the build of all transitive Dart packages & prepares code assets to
```
Comment (bkonyi) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:46]: Ubernit: s/&/and/

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 15]
```diff
@@ -30,11 +29,159 @@ import 'macos/native_assets.dart';
...
+  );
+  return nativeAssetsYamlUri;
+}
+
 /// Programmatic API to be used by Dart launchers to invoke native builds.
 ///
 /// It enables mocking `package:native_assets_builder` package.
 /// It also enables mocking native toolchain discovery via [cCompilerConfig].
-abstract class NativeAssetsBuildRunner {
+abstract class FlutterNativeAssetsBuildRunner {
```
Comment (bkonyi) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:184]: Nit: can this be an `interface class`?

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart [hunk 16]
```diff
@@ -794,10 +602,240 @@ KernelAsset _targetLocationSingleArchitecture(
...
+        fatAssetTargetLocationsIOS(codeAssets),
+        codesignIdentity,
+        buildMode,
+        fileSystem,
+      );
+    case OS.android:
+      assert(codesignIdentity == null);
+      await copyNativeCodeAssetsAndroid(
+          buildUri, assetTargetLocations, fileSystem);
+    default:
```
Comment (bkonyi) [inline packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:674]: This isn't defined in flutter_tools, but is there any reason we aren't just using enhanced enums for `OS` and `OSImpl`? This would allow for us to perform a static exhaustiveness check for all these switches without requiring us to manually raise a `StateError` everywhere.

**Review by bkonyi — APPROVED** [review]
LGTM with some nits.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 16 above)
Comment (bkonyi) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:674]: Fair enough. Should we maybe be throwing `UnimplementedError` instead then?

File: packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart (hunk 9 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:107]: I think that makes it even more verbose, because of lints that require explicit type annotations (then for both a key and a value variable).

As this is just refactoring of existing code, I hope it's ok if I leave it like this.

Comment (bkonyi) [reply packages/flutter_tools/lib/src/isolated/native_assets/macos/native_assets_host.dart:107]: Yup, that's no problem, just figured I'd mention it. I forgot about those lints...

File: packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart (hunk 12 above)
Comment (mkustermann) [reply packages/flutter_tools/test/general.shard/isolated/ios/native_assets_test.dart:194]: Yes this is supposed to be a path and the implementation greps for `iPhone Simulator` in the name. See existing code at `.../src/macos/xcode.dart:environmentTypeFromSdkroot`

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 14 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:46]: Done

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 15 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:184]: Done.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 16 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:674]: What Daco says. Adding more entries to the enum shifts from being a compile-time error to being a runtime error (and only if the new entry is used)

File: packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart (hunk 11 above)
Comment (mkustermann) [reply packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart:158]: Done.

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 13 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:37]: Done.

Though to be honest I think it makes the code worse: This class is only used internally within flutter tools. Making them named will just increase code size and reduce information density:

e.g. What used to be
```dart
return DartBuildResult(codeAssets, dependencies);
```
becomes now
```dart
return DartBuildResult(codeAssets: codeAssets, dependencies: dependencies);
```
why is the second one better than the first?

File: packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart (hunk 10 above)
Comment (mkustermann) [reply packages/flutter_tools/test/general.shard/isolated/android/native_assets_test.dart]: Done.

**Comment by mkustermann** [comment]
Thanks, @bkonyi !

File: packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart (hunk 16 above)
Comment (mkustermann) [reply packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart:674]: Done.

**Comment by eyebrowsoffire** [comment]
Reason for revert: Postsubmit failures closing the tree. See the following examples:

https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_ios%20native_assets_ios/5738/overview
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac_arm64_mokey%20native_assets_android/583/overview
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_pixel_7pro%20native_assets_android/4075/overview
https://ci.chromium.org/ui/p/flutter/builders/prod/Mac%20native_assets_ios_simulator/5287/overview

**Comment by mkustermann** [comment]
> Reason for revert: Postsubmit failures closing the tree. See the following examples:

Can reproduce the failure of the test, it was missing a case for iOS/Android, fix is
```diff
diff --git a/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart b/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart
index b28543ba16..fd8952e7d9 100644
--- a/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart
+++ b/packages/flutter_tools/lib/src/isolated/native_assets/native_assets.dart
@@ -145,7 +145,9 @@ Future<Uri?> runFlutterSpecificDartDryRunOnPlatforms({
     final OSImpl targetOS = _getNativeOSFromTargetPlatfrorm(targetPlatform);
     if (targetOS != OS.macOS &&
         targetOS != OS.windows &&
-        targetOS != OS.linux) {
+        targetOS != OS.linux &&
+        targetOS != OS.android &&
+        targetOS != OS.iOS) {
       await ensureNoNativeAssetsOrOsIsSupported(
         projectUri,
```

Make a reland CL in https://github.com/flutter/flutter/pull/155745
